lib/src/screenshots.dart

```diff
@@ -0,0 +1,327 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:pana/pana.dart';
+import 'package:pana/src/utils.dart';
+import 'package:path/path.dart' as path;
+import 'package:pubspec_parse/pubspec_parse.dart' as p;
+
+final maxFileSizeInBytes = 2097152; // 2MB
+final maxFileSizeInMegaBytes = maxFileSizeInBytes / (1024 * 1024);
+final maxNumberOfScreenshots = 5;
+
+/// [ScreenshotResult] holds either a [ProcessedScreenshot] or a list of
+/// problems describing what went wrong when trying to generate thumbnails and
+/// screenshot.
+class ScreenshotResult {
+  // The temporary directory where the processed screenshots are stored.
+  final String? _outDir;
+  ProcessedScreenshot? processedScreenshot;
+  List<String> problems;
+
+  ScreenshotResult._(this.processedScreenshot, this.problems, this._outDir);
+
+  factory ScreenshotResult.success(
+      ProcessedScreenshot processedScreenshot, String outDir) {
+    return ScreenshotResult._(processedScreenshot, [], outDir);
+  }
+
+  factory ScreenshotResult.failed(List<String> problems) {
+    return ScreenshotResult._(null, problems, null);
+  }
+
+  /// The path to where the generated WebP thumbnail is temporarily stored.
+  /// Should only be used if [processedScreenshot] is not null.
+  String get webpThumbnailPath =>
+      path.join(_outDir!, processedScreenshot!.webpThumbnail);
+
+  /// The path to where the generated PNG thumbnail is temporarily stored.
+  /// Should only be used if [processedScreenshot] is not null.
+  String get pngThumbnailPath =>
+      path.join(_outDir!, processedScreenshot!.pngThumbnail);
+
+  /// The path to where the generated WebP screenshot is temporarily stored.
+  /// Should only be used if [processedScreenshot] is not null.
+  String get webpScreenshotPath =>
+      path.join(_outDir!, processedScreenshot!.webpImage);
+}
+
+Future<List<ScreenshotResult>> processAllScreenshots(
+    List<p.Screenshot> screenshots, String pkgDir) async {
+  final results = <ScreenshotResult>[];
+  final tempDir = Directory.systemTemp.createTempSync().path;
```

Comment:
Are we going to delete this temporary folder? When?
Could we do this in a ` finally {...}` block?

Reply:
Maybe, we should just make `ScreenshotResult.webpThumbnail` a `Uint8List` as proposed above... That way we can delete this temporary directory at the end of `processAllScreenshots` in a finally-block.

This assumes we can reasonably keep all screenshots in memory? But don't we have sensible size limits that ensures this won't be a problem?

Note. I'm proposing that after generating the image, we read the image file and keep it memory as `Uint8List`.
That way we can delete the temporary folder when this function returns.

Reply:
Done.